src/behaviors.ts

```diff
@@ -179,18 +188,17 @@ const behaviorAvailable = <TArgs extends unknown[]>(
   )
 }
 
-const behaviorMatches = <TArgs extends unknown[]>(args: TArgs) => {
+const behaviorMatches = <TArgs extends unknown[]>(actualArguments: TArgs) => {
   return (behavior: BehaviorEntry<TArgs>): boolean => {
-    let index = 0
-
-    while (index < args.length || index < behavior.args.length) {
-      if (!equals(args[index], behavior.args[index])) {
-        return false
-      }
-
-      index += 1
-    }
-
-    return true
+    // Check arity
+    const expectedArguments = behavior.args
+    const { ignoreExtraArgs } = behavior
+    if (expectedArguments.length !== actualArguments.length && !ignoreExtraArgs)
+      return false
+
+    // Check arguments
+    return expectedArguments.every((expectedArgument, index) => {
+      return equals(actualArguments[index], expectedArgument)
+    })
```

Comment:
I like this arity check before the arguments - it makes me realize there's a subtle bug in the current implementation around explicit vs implicit `undefined` values

A few suggestions, in priority order

- Should the check be `>` in the `ignoreExtraArgs` case?
- Let's keep this as a regular loop to avoid the extra function creation
- I don't think these comments add any extra value to the code, let's lose them

```suggestion
    const { args: expectedArguments, ignoreExtraArgs } = behavior

    if (
      expectedArguments.length !== actualArguments.length ||
      (ignoreExtraArgs && expectedArguments.length > actualArguments.length)
    ) {
      return false
    }

    return expectedArguments.every((expectedArgument, index) => {
      return equals(actualArguments[index], expectedArgument)
    })
```

Reply:
Given the following specification of `ignoreExtraArgs`...

> Check all arguments in `calledWith` to find a match, but past the arguments specified in `calledWith`, ignore any extra arguments in the call itself when determining the match

...with`ignoreExtraArgs` set, what do you expect the following stubbing to do?

```ts
const spy = when(vi.fn(), { ignoreExtraArgs: true })
  .calledWith(undefined)
  .thenReturn('called')

const result = spy() // should this be `called` or `undefined`?
```

My gut says `when(vi.fn(), { ignoreExtraArgs: true }).calledWith(undefined)` should **not** match `spy()`. However, with a `!==` check, this _does_ match.

Passing `undefined` explicitly is _extremely_ similar to, but not technically exactly the same as, omitting an argument, so I think if the user says "I expect an explicit `undefined`" it should be respected, even if `ignoreExtraArgs` is set